### DISCUSSION TITLE
traces/tui: 'exec' should be an event in withExec span, not a separate span

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12286] [maintainers]
## Problem

- When `Container.withExec` is called, it emits an otel span. When it returns a (lazy) container ID, that span ends.
- Later, this may cause an actual `exec`, with a *separate* span
- These two operations are highly coupled, but having them in separate spans does not convey that
- The current stopgap of "transplanting" the exec span as a child of the `withExec` span is an improvement, but it's not enough
- By showing to the user a `withExec` span and a separate `exec` span, we imply that these two operations are of the same nature (they are not), are both documented (they are not), and can both be used and reasoned about in the same way (they can't).
- On top of the confusion caused by erroneously placing these two operations on the same plane, there is also cognitive overload. Traces are already pretty busy. This is one more level of hierarchy to navigate. In large traces, it adds up - especially when searching for the source of an error.

<img width="1274" alt="Screenshot 2024-07-17 at 1 52 22 PM" src="https://github.com/user-attachments/assets/57b245fd-7285-42a6-8c6c-86938815db0e">

## Solution

- Merge `exec` span into the `withExec` span. They are in fact the same operation.
- Make `exec` an event within the combined span.
- This implies that `withExec` span does not end when initial API request is completed. That becomes an event too.

**Comment by shykes** [comment]
FYI @vito @crjm @aluzzardi

**Comment by crjm** [comment]
We **could** get rid of it, are we sure we want to hide to the user the internal distinction between `dagql` and the operation being performed?
I agree it improves readability while scrolling, but it might help debugging when trying to look for a specific error?

**Comment by shykes** [comment]
I could be wrong, but I think the internal distinction will still be available in the form of events. Normally there shouldn't be any loss of information - just a slight reframing of the same information.

**Comment by vito** [comment]
> This implies that `withExec` span does not end when initial API request is completed. That becomes an event too.

Wouldn't this arbitrarily inflate the duration of the call? (Time between `withExec` call and the actual `exec` starting)

**Comment by vito** [comment]
Here's a proposal:

* Move logs and spans to from `exec foo` to `withExec` (or `asService` for services).
  * This part is easy; we already do it for functions.
* Keep the separate effect spans, but as an internal implementation detail.
  * That way we can still distinguish the duration of the synchronous part (`withExec`) and the asynchronous parts (`exec foo`).
* In the web UI, don't show effect spans in the tree - only show them in the timeline, styled the same as the original span's bar, maybe with a tooltip saying which effect it is (e.g. `exec foo`).
  * In the TUI, there's no timeline, but the duration already reflects 'active time' so no change needed.
* In the UI, distinguish between a `withExec` whose synchronous part has completed and a `withExec` whose effects have completed.
  * Currently shows green check immediately which feels misleading. Can dim it, or give it a separate status glyph.

I think that addresses the original request while maintaining the integrity of our data.
